plugin/internal/remotecall/remotecall.go

```diff
@@ -0,0 +1,235 @@
+	if err != nil {
+		return nil, fmt.Errorf("error reading mapping key: %w", err)
+	}
+
+	return &RemoteCall{
+		userClusterNameURL:                   userClusterNameURLParsed,
+		userClusterToken:                     userClusterToken,
+		clusterNameInClusterAuthenticatorURL: clusterNameInClusterAuthenticatorURLParsed,
+		serviceJwtProvider:                   jwtProviders,
+		httpClient:                           createHttpClient(),
+		mappingKeyFile:                       key,
+		clusterNameUpstreamURL:               clusterNameUpstreamURL,
+	}, nil
+}
+
+func createHttpClient() *http.Client {
+	return &http.Client{
+		Timeout: 10 * time.Second,
+		Transport: &http.Transport{
+			MaxIdleConns:    10,
+			IdleConnTimeout: 30 * time.Second,
+		},
+	}
+}
+
+func (r *RemoteCall) GetClusterName(username string) (string, error) {
+	if username == "" {
+		return "", fmt.Errorf("empty username")
+	}
+	req, err := http.NewRequest("GET", r.userClusterNameURL.JoinPath(username).String(), nil)
+	if err != nil {
+		return "", fmt.Errorf("error creating request for getting cluster name with url: %q and username:"+
+			" %q :%w", r.userClusterNameURL, username, err)
+	}
+
+	// Set custom headers if needed
+	req.Header.Set(UserAgent, UserAgentSSHGateway)
+	req.Header.Set(Accept, ApplicationJson)
+	req.Header.Set(ContentType, ApplicationJson)
+	req.Header.Set(AuthTokenUserClusterMapping, r.userClusterToken)
+
+	resp, err := r.httpClient.Do(req)
+	if err != nil {
+		return "", fmt.Errorf("error making request: %w", err)
+	}
+	if resp != nil && resp.Body != nil {
+		defer func() { _ = resp.Body.Close() }()
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		bodyBytes, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return "", fmt.Errorf("error reading response body during GetClusterName: %w", err)
+		}
+		return "", fmt.Errorf("error: status code for url: %s, error: %q: %d", req.URL.String(), bodyBytes,
+			resp.StatusCode)
+	}
+
+	userClusterResponse := UserClusterResponse{}
+	err = json.NewDecoder(resp.Body).Decode(&userClusterResponse)
+	if err != nil {
+		return "", fmt.Errorf("error reading response body for GetClusterName: %w", err)
+	}
+
+	return userClusterResponse.ClusterName, nil
+}
+
+// todo: refactor once we have user info in directory svc
+
+func (r *RemoteCall) AuthenticateKey(
+	key []byte,
+	keyType string,
+	clusterURL string,
+	clusterName string,
+	accountId string,
+) (*UserKeyAuthResponse, error) {
+	token, err := r.getUpstreamAuthenticatorAuthToken(clusterName)
+	if err != nil {
+		return nil, fmt.Errorf("error getting authenticator token from cluster name: %w", err)
+	}
+
+	sshKeyObj := sshKeyObject{
+		Key:       key,
+		Algorithm: keyType,
+	}
+	auth := userKeyAuthRequest{
+		SshKeyObject: sshKeyObj,
+		AccountId:    accountId,
+	}
+	body, err := json.Marshal(auth)
+	if err != nil {
+		return nil, fmt.Errorf("error marshalling auth: %v", auth)
+	}
+
+	req, err := http.NewRequest("POST", clusterURL, bytes.NewBuffer(body))
```

Comment:
I see many people use bytes.NewBuffer is there any reason to use mutable buffer for read operation. I almost always use bytes.NewReader just to read data.

Reply:
good point!